soda/cmd/validate/model.go

```diff
@@ -0,0 +1,111 @@
+package validate
+
+import (
+	"fmt"
+	"go/ast"
+	"regexp"
+	"strings"
+)
+
+type model struct {
+	packages map[string]*ast.Package
+	tags     map[string][]string
+	rules    []Checker
+}
+
+type Checker interface {
+	Check(tag string, structName string) []ValidationError
+}
+
+type ValidationError struct {
+	invalidSymbols string
+	field          string
+	structName     string
+	duplicate      bool
+}
+
+func (e *ValidationError) Error() string {
+	return fmt.Sprintf("Invalid symbols '%v' contained in %v.%v", e.invalidSymbols, e.structName, e.field)
+}
+
+type Rule struct {
+	expr *regexp.Regexp
+}
+
+func (r *Rule) Check(tag string, structName string) []ValidationError {
+	match := r.expr.FindString(tag)
+	errorss := []ValidationError{}
+
+	if len(match) > 0 {
+		err := ValidationError{match, tag, structName, false}
+		errorss = append(errorss, err)
+	}
+
+	return errorss
+}
+
+func checkForDuplicates(tag string, structName string, fieldsCache map[string]bool) []ValidationError {
+	errorss := []ValidationError{}
+	cacheKey := strings.Join([]string{structName, tag}, ".")
+
+	if _, exist := fieldsCache[cacheKey]; exist {
+		err := ValidationError{"duplicate entry", tag, structName, true}
+		errorss = append(errorss, err)
+	}
+
+	fieldsCache[cacheKey] = true
+
+	return errorss
+}
+
+func NewModel() model {
+	m := model{}
+	m.packages = getPackages("models")
+	m.tags = getTags("db", m.packages)
```

Comment:
what about the other tags, like `has_many` or `belongs_to`? those should also be validated, but with probably a different set of rules.